pkg/rosa/client.go

```diff
@@ -1,27 +1,56 @@
 package rosa
 
 import (
+	"context"
 	"fmt"
 	"os"
 
 	sdk "github.com/openshift-online/ocm-sdk-go"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api-provider-aws/v2/pkg/cloud/scope"
+)
+
+const (
+	ocmTokenKey  = "ocmToken"
+	ocmAPIURLKey = "ocmApiUrl"
 )
 
 type rosaClient struct {
-	ocm *sdk.Connection
+	ocm       *sdk.Connection
+	rosaScope *scope.ROSAControlPlaneScope
 }
 
-// NewRosaClientWithConnection creates a client with a preexisting connection for testing purpose
-func NewRosaClientWithConnection(connection *sdk.Connection) *rosaClient {
+// NewRosaClientWithConnection creates a client with a preexisting connection for testing purposes.
+func NewRosaClientWithConnection(connection *sdk.Connection, rosaScope *scope.ROSAControlPlaneScope) *rosaClient {
 	return &rosaClient{
-		ocm: connection,
+		ocm:       connection,
+		rosaScope: rosaScope,
 	}
 }
 
-func NewRosaClient(token string) (*rosaClient, error) {
-	ocmAPIUrl := os.Getenv("OCM_API_URL")
-	if ocmAPIUrl == "" {
-		ocmAPIUrl = "https://api.openshift.com"
+func NewRosaClient(ctx context.Context, rosaScope *scope.ROSAControlPlaneScope) (*rosaClient, error) {
```

Comment:
Why are we doing the lookup inside here instead of loading config at startup and using it as needed?

Do we need to handle re-loading this value if the Secret changes?

Reply:
the secret is referenced through the API, so we can only read it in the reconciliation loop.
Also its content could be changed, so this ensures we always fetch the latest version.

Reply:
Are we re-loading this and re-creating the client + connections on every single loop? Wouldn't we prefer to load it once and keep track of changes? inotify watch should suffice

Reply:
I think watching for secret content changes produces more overhead and code complexity, given that reading the secret every loop is not really a problem with caching enabled.

We would also need a different mechanism to detect if `.spec.CredentialsSecreRef` has changed.
It is possible that `.spec.CredentialsSecreRef` has been updated in the new reconciliation loop, so we need to keep reading the field and fetch the secret.

Reply:
That seems super wasteful from a resources standpoint - open connections are costly. What's our expected rate of change here?

Reply:
If `rosaScope.Client.Get` is using a lister cache, this point is moot and we can iterate on it if we see a performance problem. Many Kubernetes controllers use custom caches (see the PKI caches in core, or every single controller using a lister), so I don't know how dogmatic the stance is on in-memory state between loops. :)